concourse/client/util.py

```diff
@@ -122,3 +137,118 @@ def jobs_not_triggered(
+                )
+
+                if wait_for_job_to_be_triggered(
+                    job=job,
+                    resource_version=resource_version,
+                    concourse_api=concourse_api,
+                    retries=10,
+                    sleep_time_seconds=5
+                ):
+                    info(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        'has been triggered'
+                    )
+                else:
+                    warning(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        'could not be triggered'
+                    )
+                return
+        except PinningFailed:
+            retries -= 1
+
+    warning(f'Could not process untriggered job {job.name}')
+
+
+@contextmanager
+def pin_resource(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    pin_comment = {
+        'datetime': datetime.now().isoformat(),
+        'resource_type': resource.type().value,
+        'version_data': resource_version.version(),
+        'comment': 'automatically pinned by technical user, please do not unpin manually',
+    }
+    _ensure_resource_unpinned(
+        resource=resource,
+        concourse_api=concourse_api,
+        retries=10,
+        sleep_time_seconds=5,
+    )
+    try:
+        concourse_api.pin_and_comment_resource_version(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+            resource_version_id=resource_version,
+            comment=json.dumps(pin_comment),
+        )
+        yield
+    finally:
+        concourse_api.unpin_resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+
+def _ensure_resource_unpinned(
+    resource: PipelineResource,
+    concourse_api,
+    retries: int=10,
+    sleep_time_seconds: int=5,
+):
+    while retries >= 0:
+        cc_resource = concourse_api.resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+        if not cc_resource.is_pinned():
+            return
+
+        pin_comment = cc_resource.pin_comment()
+        try:
+            # if comment is json format, we assume it was pinned by us
+            parsed_comment = json.loads(pin_comment)
+            pin_time = datetime.fromisoformat(parsed_comment['datetime'])
+            if pin_time + timedelta(minutes=3) < datetime.now():
+                # resource pinned by us but unpin did not happen for more than 3 minutes -> cleanup
+                concourse_api.unpin_resource(
+                    pipeline_name=resource.pipeline_name(),
+                    resource_name=resource.name,
+                )
+                return
+        except:
+            # resource pinned by someone else, wait and try again
+            pass
```

Comment:
This catches _all_ raised errors. 
```suggestion
        except JSONDecodeError:
            # resource pinned by someone else, wait and try again
            pass
```
As for handling other errors: Not sure. I think it is already retrying on connection errors, so that should be fine.